src/ucp/wireup/select.c

```diff
@@ -1566,6 +1595,79 @@ ucp_wireup_select_params_init(ucp_wireup_select_params_t *select_params,
     select_params->show_error    = show_error;
 }
 
+static double
+ucp_wireup_keepalive_score_func(ucp_worker_h worker, ucp_rsc_index_t rsc_index,
+                                const uct_md_attr_t *md_attr,
+                                const uct_iface_attr_t *iface_attr,
+                                const ucp_address_iface_attr_t *remote_iface_attr)
+{
+    ucp_context_h context                  = worker->context;
+    ucp_worker_iface_t *wiface             =
+            ucp_worker_iface(worker, rsc_index);
+    uct_perf_attr_t perf_attr              = {
+        .field_mask = UCT_PERF_ATTR_FIELD_MAX_INFLIGHT_EPS
+    };
+    const uct_tl_resource_desc_t *resource =
+            &context->tl_rscs[rsc_index].tl_rsc;
+    ucs_status_t status;
+
+    status = uct_iface_estimate_perf(wiface->iface, &perf_attr);
+    if (status != UCS_OK) {
+        ucs_warn(UCT_TL_RESOURCE_DESC_FMT
+                 ": getting perf estimations failed: %s",
+                 UCT_TL_RESOURCE_DESC_ARG(resource),
+                 ucs_status_string(status));
+        return 0;
+    }
+
+    return ucp_wireup_am_score_func(worker, rsc_index, md_attr, iface_attr,
+                                    remote_iface_attr) *
+           (perf_attr.max_inflight_eps / SIZE_MAX);
+}
+
+/* Lane for doing keepalive */
+static ucs_status_t
+ucp_wireup_add_keepalive_lane(const ucp_wireup_select_params_t *select_params,
+                              ucp_err_handling_mode_t err_mode,
+                              ucp_wireup_select_context_t *select_ctx)
+{
+    ucp_ep_h ep                          = select_params->ep;
+    ucp_worker_h worker                  = ep->worker;
+    ucp_wireup_select_info_t select_info = {0};
+    unsigned ep_init_flags               = ucp_wireup_ep_init_flags(
+                                                   select_params, select_ctx);
+    ucp_wireup_criteria_t criteria;
+    ucs_status_t status;
+
+    if ((err_mode == UCP_ERR_HANDLING_MODE_NONE) ||
+        !ucp_worker_keepalive_is_enabled(worker) || 
```

Comment:
why are we checking these two conditions again in ucp_wireup_check_keepalive lines 259-260?
imo they should be only here, and removed from ucp_wireup_check_keepalive

Reply:
good catch, remove uneeded checks from `ucp_wireup_check_keepalive`